opentelemetry/api/trace/__init__.py

```diff
@@ -0,0 +1,241 @@
+        """Gets the currently active span from the context.
+
+        If there is no current span, return a placeholder span with an invalid
+        context.
+
+        Returns:
+            The currently active :class:`.Span`, or a placeholder span with an
+            invalid :class:`.SpanContext`.
+        """
+        raise NotImplementedError
+
+
+    @contextmanager
+    def start_span(self, name: str, parent: Span) -> Iterator[Span]:
+        """Context manager for span creation.
+
+        Create a new child of the current span, or create a root span if no
+        current span exists. Start the span and set it as the current span in
+        this tracer's context.
+
+        On exiting the context manager stop the span and set its parent as the
+        current span.
+
+        Example::
+
+            with tracer.span("one") as parent:
+                parent.add_event("parent's event")
+                with tracer.span("two") as child:
+                    child.add_event("child's event")
+                    tracer.get_current_span()  # returns child
+                tracer.get_current_span()      # returns parent
+            tracer.get_current_span()          # returns the previously active span
+
+        This is a convenience method for creating spans attached to the
+        tracer's context. Applications that need more control over the span
+        lifetime should use :meth:`create_span` instead. For example::
+
+            with tracer.start_span(name) as span:
+                do_work()
+
+        is equivalent to::
+
+            span = tracer.create_span(name, parent=tracer.get_current_span())
+            with tracer.use_span(span):
+                do_work()
+
+        Args:
+            name: The name of the span to be created.
+            parent: The span's parent.
+
+        Yields:
+            The newly-created span.
+        """
+        raise NotImplementedError
+
+    def create_span(self, name: str, parent: Span) -> Span:
+        """Creates a new child span of the given parent.
+
+        Creating the span does not start it, and should not affect the tracer's
+        context. To start the span and update the tracer's context to make it
+        the currently active span, see :meth:`use_span`.
+
+        Applications that need to explicitly set spans' parents or create spans
+        detached from the tracer's context should use this method.
+
+            with tracer.start_span(name) as span:
+                do_work()
+
+        This is equivalent to::
+
+            span = tracer.create_span(name, parent=tracer.get_current_span())
+            with tracer.use_span(span):
+                do_work()
+
+        Args:
+            name: The name of the span to be created.
+            parent: The span's parent.
+
+        Returns:
+            The newly-created span.
+        """
+        raise NotImplementedError
+
+    @contextmanager
+    def use_span(self, span: Span) -> Iterator[None]:
```

Comment:
For languages which don't have the `with` or `using` concept, it might be good to have a dedicated API. For Python, `with span` (`Span.__enter__` and `Span.__exit__`) could be a good replacement for `use_span`?

Reply:
FYI @c24t 

https://github.com/census-instrumentation/opencensus-python/blob/master/context/opencensus-context/examples/async_span.py

https://github.com/census-instrumentation/opencensus-python/blob/master/context/opencensus-context/examples/span.py

Reply:
I agree if we can find a nice way to handle modifying the context.